liblist.c

```diff
@@ -156,16 +137,11 @@ void *ListRemoveEnd(List *l) {
     return NULL;
 }
 
-/*
- * Remove o elemento 'elemento' caso ele exista na lista.
- * Retorna 1 se a operacao foi bem sucedida e 0 caso contrario.
- * Se o elemento nao for encontrado na lista tambem retorna 0.
- */
-void *ListRemoveKey(List *l, void *key) {
-    if (!ListIsEmpty(l) && ListContains(l, key)) {
+void *ListSoftRemoveKey(List *l, void *key, int (*comparator)(void *, void *)) {
+    if (!ListIsEmpty(l) && ListFindKey(l, key, comparator)) {
         Node *target = l->head;
 
-        while (target->key != key) {
+        while (!(comparator(target->key, key))) {
             target = target->next;
         }
 
```

Comment:
Se target for o último elemento da lista, não vai dar seg fault dele tentando acessar target->next?!

Reply:
- Em caso do target ser o último elemento da lista, o acesso à target->next nas linhas 149 e 151 irá retornar NULL. Irá atribuir NULL à target->prev->next ou a l->head, mas não dá segFault visto que não tenta-se acessar nenhum atributo dele (NULL->key daria problema);
- Já na linha 155 eu acesso um atributo do target->next, onde poderia dar segFault caso o mesmo fosse NULL, por isso fez-se necessária a verificação da linha 154: if (target->next) equivale à if (target->next!=NULL);

Reply:
Truee, real...